app/services/email_sender/notify_sender.rb

```diff
@@ -0,0 +1,37 @@
+require "notifications/client"
+
+module EmailSender
+  class NotifySender
+    def call(options)
```

Comment:
Can we use `def call(address:, template_id:)` here instead of having to `unpack` and check that we were given a hash?

Reply:
so the reason I've left this without keyword args for the moment is because it is a minimal interface. We will most probably want to add more than just address and template....things like our reference to pass through to notify, the subject, perhaps a user's name for personalisation. It felt premature to start future proofing that right now given we are creating the MVP.

Those may or may not be good reasons for the decision though?

Reply:
Hmm, yeah, I can see where you're coming from.

I would argue though, that we're going to need to modify the code anyway to use any new fields from the `options` hash that it doesn't give us any advantages here.

Reply:
Could the list of arguments get quite large when we start to add new features? That would be a good argument for using `options` and unpacking it; it's also then easier to have a mix of required and optional params in any order.

Reply:
Required and optional params can be achieved with `def func(a:, b: "optional")` and the arguments can be written in any order when calling the method.

I don't like using `options` because it makes it much harder to immediately see what arguments a method takes without writing documentation for the method, plus it means Ruby gives us a load of `ArgumentError`s for free.

Reply:
In the first instance, the list of arguments could be as large as the list of arguments to notify https://github.com/alphagov/notifications-ruby-client#arguments:

From their example:
```
  email_address: email_address,
  template_id: template_id,
  personalisation: {
    name: "name",
    year: "2016"
  },
  reference: "your_reference_string"
  email_reply_to_id: email_reply_to_id
```

Personalisation is the bit that can get really large since that's where we'll pass in any template based variables.

Reply:
```ruby
def call(email_address:, template_id:, personalisation: {}, reference: nil, email_reply_to_id: nil)
```

Personally that seems fine to me, we can always break it down into lines if we want to. The extra documentation we get from doing it like that seems like a big win to me.

Happy to concede on this point though if I'm the only advocate for this!

Reply:
I'm tinkering with turning the function definition into

```
def call(address:, **keyword_args)
```

that should give us the required `ArgumentError` for free on `address` as @thomasleese points out and allows us to pass in any other number of keyword arguments. If we decide to explicitly require another argument in the future, we can introduce the requisite argument in any consumers first without breaking anything and then pull it out specifically in the function definition.

Reply:
@bevanloon That seems like a good compromise! 👍